File: app/src/main/java/com/example/myapplication/core/nutrition/NutritionScoreCalculator.kt
package com.example.myapplication.core.nutrition

import kotlin.math.abs
import kotlin.math.min
import kotlin.math.roundToInt

object NutritionScoreCalculator {
    /**
     * Calculates a daily nutrition score from 0 to 100 based on calories, macros, and water intake.
     */
    fun calculateScore(
        consumed: Nutrients,
        target: NutritionTarget?,
        waterIntakeMl: Int,
        waterTargetMl: Int = 2000
    ): CalculationResult {
        if (target == null || target.calories <= 0) {
            return CalculationResult(0, "Chưa thiết lập mục tiêu", "🔴")
        }

        // Calories: 30 points max
        val calorieScore = if (target.calories > 0) {
            val deviationRatio = abs(consumed.calories - target.calories).toDouble() / target.calories
            (30.0 * (1.0 - deviationRatio).coerceIn(0.0, 1.0)).roundToInt()
        } else 0

        // Protein: 25 points max
        val proteinScore = if (target.proteinGrams > 0) {
            val ratio = consumed.proteinGrams.toDouble() / target.proteinGrams
            (25.0 * min(1.0, ratio)).roundToInt()
        } else 0

        // Carbs: 15 points max
        val carbsScore = if (target.carbsGrams > 0) {
            val deviationRatio = abs(consumed.carbsGrams - target.carbsGrams).toDouble() / target.carbsGrams
            (15.0 * (1.0 - deviationRatio).coerceIn(0.0, 1.0)).roundToInt()
        } else 0

        // Fat: 15 points max
        val fatScore = if (target.fatGrams > 0) {
            val deviationRatio = abs(consumed.fatGrams - target.fatGrams).toDouble() / target.fatGrams
            (15.0 * (1.0 - deviationRatio).coerceIn(0.0, 1.0)).roundToInt()
        } else 0

        // Water: 15 points max
        val waterScore = if (waterTargetMl > 0) {
            val ratio = waterIntakeMl.toDouble() / waterTargetMl
            (15.0 * min(1.0, ratio)).roundToInt()
        } else 0

        val totalScore = calorieScore + proteinScore + carbsScore + fatScore + waterScore

        val labelAndEmoji = when {
            totalScore >= 90 -> Pair("Xuất sắc", "🌟")
            totalScore >= 70 -> Pair("Tốt", "✅")
            totalScore >= 50 -> Pair("Cần cải thiện", "⚠️")
            else -> Pair("Chưa đạt", "🔴")
        }

        return CalculationResult(totalScore, labelAndEmoji.first, labelAndEmoji.second)
    }

    data class CalculationResult(
        val score: Int,
        val label: String,
        val emoji: String
    )
}

File: flutter/lib/core/nutrition/nutrition_score_calculator.dart
import 'dart:math';
import '../model/nutrition_models.dart';

class NutritionScoreResult {
  final int score;
  final String label;
  final String emoji;

  NutritionScoreResult({
    required this.score,
    required this.label,
    required this.emoji,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is NutritionScoreResult &&
          runtimeType == other.runtimeType &&
          score == other.score &&
          label == other.label &&
          emoji == other.emoji;

  @override
  int get hashCode => score.hashCode ^ label.hashCode ^ emoji.hashCode;
}

class NutritionScoreCalculator {
  static NutritionScoreResult calculateScore({
    required Nutrients consumed,
    NutritionTarget? target,
    required int waterIntakeMl,
    int waterTargetMl = 2000,
  }) {
    if (target == null || target.calories <= 0) {
      return NutritionScoreResult(
        score: 0,
        label: "Chưa thiết lập mục tiêu",
        emoji: "🔴",
      );
    }

    // Calories: tối đa 30 điểm
    final calorieScore = target.calories > 0
        ? (30.0 *
                (1.0 -
                        (consumed.calories - target.calories).abs() /
                            target.calories)
                    .clamp(0.0, 1.0))
            .round()
        : 0;

    // Protein: tối đa 25 điểm
    final proteinScore = target.proteinGrams > 0
        ? (25.0 * min(1.0, consumed.proteinGrams / target.proteinGrams)).round()
        : 0;

    // Carbs: tối đa 15 điểm
    final carbsScore = target.carbsGrams > 0
        ? (15.0 *
                (1.0 -
                        (consumed.carbsGrams - target.carbsGrams).abs() /
                            target.carbsGrams)
                    .clamp(0.0, 1.0))
            .round()
        : 0;

    // Fat: tối đa 15 điểm
    final fatScore = target.fatGrams > 0
        ? (15.0 *
                (1.0 -
                        (consumed.fatGrams - target.fatGrams).abs() /
                            target.fatGrams)
                    .clamp(0.0, 1.0))
            .round()
        : 0;

    // Water: tối đa 15 điểm
    final waterScore = waterTargetMl > 0
        ? (15.0 * min(1.0, waterIntakeMl / waterTargetMl)).round()
        : 0;

    final totalScore = calorieScore + proteinScore + carbsScore + fatScore + waterScore;

    final String label;
    final String emoji;
    if (totalScore >= 90) {
      label = "Xuất sắc";
      emoji = "🌟";
    } else if (totalScore >= 70) {
      label = "Tốt";
      emoji = "✅";
    } else if (totalScore >= 50) {
      label = "Cần cải thiện";
      emoji = "⚠️";
    } else {
      label = "Chưa đạt";
      emoji = "🔴";
    }

    return NutritionScoreResult(
      score: totalScore,
      label: label,
      emoji: emoji,
    );
  }
}
